[Rebecca Lau] iP

rebeccalaujx opened the pull request:
### Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))


Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/rebeccalaujx/ip/).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X]  Managing tasks
- [ ]  Managing deadlines
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
    public static void main(String[] args) {
        new Duke("duke.txt").run();
    }
```

Commit:
Add Gradle support

rebeccalaujx committed:
Level-1

rebeccalaujx committed:
Level-2

rebeccalaujx committed:
Level-3

rebeccalaujx committed:
Level-3

rebeccalaujx committed:
Level-4

rebeccalaujx committed:
A-TextUiTesting

rebeccalaujx committed:
A-TextUiTesting

rebeccalaujx committed:
Level-5

rebeccalaujx committed:
Level-5

rebeccalaujx committed:
Level-6

rebeccalaujx committed:
save tasks to hard disk

rebeccalaujx committed:
changes dates into a MMM dd yyyy format

rebeccalaujx committed:
save tasks to disk

rebeccalaujx committed:
merged branch 7

rebeccalaujx committed:
change dates to MMM dd yyyy format

rebeccalaujx committed:
merged branch 8

rebeccalaujx committed:
add more OOP

rebeccalaujx committed:
added some OOP

rebeccalaujx committed:
added more oop

rebeccalaujx committed:
add files to packages

kleonang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,25 @@
+package duke;
+
+import duke.command.*;
```
Would be better to specifically list all `duke.command` imports here instead of using wildcard imports

kleonang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,25 @@
+            return new ExitCommand();
+        } else if (input.equals("list")) {
+            return new ListCommand();
+        } else if (input.startsWith("done")) {
+            return new DoneCommand(input);
+        } else if (input.startsWith("delete")) {
+            return new DeleteCommand(input);
+        } else if (input.startsWith("todo")) {
+            return new TodoCommand(input);
+        } else if (input.startsWith("deadline")) {
+            return new DeadlineCommand(input);
+        } else if (input.startsWith("event")) {
+            return new EventCommand(input);
```
Using `input.startsWith("...")` may cause some invalid commands to be interpreted as valid commands, e.g. when input starts with `deadliner` or `todoy`. Instead, consider using `input.split(" ", 2)` to get a `String[]`, then check equality against the first element of the `String[]`. This way, the second element of the `String[]` also contains all the information you need for your `Command` subclasses later on.

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+package duke;
+
+import duke.task.*;
```
Would be better to specifically list imports as before

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Scanner;
+
+public class Storage {
+    String filePath;
+    PrintWriter writer;
+    TaskList ls;
```
Would be good to add `private` access modifiers here

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+        try {
+            FileWriter fw = new FileWriter(filePath, false);
+            writer = new PrintWriter(fw);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Possible to combine into a single catch statement here
```suggestion
        } catch (FileNotFoundException | IOException e) {
            e.printStackTrace();
        } 
```

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public TaskList load() throws IOException, DukeException {
+        TaskList ls = new TaskList();
```
Instead of `ls`, a more readable variable name for your TaskList here could be `tasks`.

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+            String taskDate = getDate(input);
+            Event eTask = new Event(taskDesc, taskDate);
+            return eTask;
+        }
```
Could potentially improve readability by renaming `tTask`, `dTask` and `eTask` into `todo`, `deadline` and `event` respectively.

kleonang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+                }
+            }
+        }
+        return input.substring(endIndex);
```
Could potentially shorten code by using `input.split("|")` to get a `String[]`, then return the last element of the `String[]`.

kleonang commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,66 @@
+        System.out.println("Noted. I've removed this task: \n" +
+                taskToString
+                + "\nNow you have " + size + " tasks in the list.");
+    }
```
It's possible to just use `task` here as the conversion to String type is done implicitly (like how you don't have to convert `size` to a String).
```suggestion
    public void addTaskToList(Task task, int size) {
        System.out.println("Got it. I've added this task: \n" + task
                + "\nNow you have " + size + " tasks in the list.");
    }

    public void removeTaskFromList(Task task, int size) {
        System.out.println("Noted. I've removed this task: \n" +
                task
                + "\nNow you have " + size + " tasks in the list.");
    }
```

kleonang commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,37 @@
+    public DeleteCommand(String input) throws DukeException {
+        this.input = input;
+        try {
+            if (input.equals("delete") || input.equals("delete ")) {
```
This can be avoided if the `input.split(" ", 2)` method is used in `Parser.java`, in which case you can just check if the second element of the `String[]` is empty.

kleonang commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,10 @@
+
+public abstract class Command {
+    public abstract void execute(TaskList ls, Ui ui, Storage storage) throws DukeException;
+    public abstract boolean isExit();
```
Since all but one of the commands have `isExit()` as `false`, this can be made into a non-abstract method.
```suggestion
    public boolean isExit() {
        return false;
    }
```

kleonang commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,34 @@
+        storage.rewriteFile(ls);
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
After the non-abstract `isExit()` is implemented in `Command.java`, there's no need to override here (and all commands except `ExitCommand.java`) anymore.

kleonang commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,18 @@
+        System.exit(0);
+    }
+
+    @Override
+    public boolean isExit() {
+        return true;
+    }
```
Here should be the only `Command` subclass where overriding the `isExit()` method is necessary after implementing it in `Command.java`.

kleonang commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+            throw new DukeException("duke.task.Deadline should be in a yyyy-mm-dd format.");
+        }
+
+        if (description.isEmpty() || description == "" || description == " ") {
+            throw new DukeException("☹ OOPS!!! The description of a deadline cannot be empty.");
+        } else {
+            this.description = description.substring(1);
+        }
+
+        if (deadline.isEmpty() || deadline == "" || deadline == " ") {
+            throw new DukeException("☹ OOPS!!! The deadline of this task must be indicated.");
```
As before, these repeated checks in the if statements should not be necessary after using `input.split(" ", 2)` in `Parser.java`.

kleonang commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+            throw new DukeException("duke.task.Event date should be in a yyyy-mm-dd format.");
+        }
+
+        if (description.isEmpty() || description == "" || description == " ") {
+            throw new DukeException("☹ OOPS!!! The description of an event cannot be empty.");
+        } else {
+            this.description = description.substring(1);
+        }
+
+        if (at.isEmpty() || at == "" || at == " ") {
+            throw new DukeException("☹ OOPS!!! The time of the event must be indicated.");
```
As before, these repeated checks in the if statements should not be necessary after using `input.split(" ", 2)` in `Parser.java`.

kleonang reviewed (commented):
Hi there, I've looked through your code and your code quality is generally good with some minor improvements to be made 😊. 

Good use of OOP is evident, especially through the abstract `Task` and `Command` classes. You may also consider adding JavaDocs to the public constructors and methods to ease readability. 

Great work on the code overall! 👍

rebeccalaujx committed:
added JUnit tests

rebeccalaujx committed:
packaged the app as an executable JAR file

rebeccalaujx committed:
added JavaDocs

rebeccalaujx committed:
Tweaked the code to comply with a coding standard

rebeccalaujx committed:
Gave users a way to find a task by searching for a keyword

rebeccalaujx committed:
Merge branch 'branch-A-JavaDoc' and 'branch-A-CodingStandard'

rebeccalaujx committed:
Merge branch 'branch-Level-9'

HangZelin commented on src/main/java/duke/Parser.java:
```diff
@@ -1,3 +1,7 @@
+package duke;
+
+import duke.command.*;
```
Is it better to not use wildcard for the import? It might be better to list all the stuff you want to import.

HangZelin commented on src/main/java/duke/Storage.java:
```diff
@@ -1,9 +1,12 @@
+package duke;
+
+import duke.task.*;
```
It might be better to list all the stuff you want to import instead of using wildcard.

HangZelin commented on src/main/java/duke/Ui.java:
```diff
@@ -15,7 +20,7 @@ public void showWelcome() {
                 + "|____/ \\__,_|_|\\_\\___|\n";
         System.out.println("Hello from\n" + logo);
-        System.out.println("Hello! I'm Duke\n" + "What can I do for you?");
+        System.out.println("Hello! I'm duke.Duke\n" + "What can I do for you?");
```
You may import the duke package inside, and you not really have to specify package all the time.

HangZelin commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -14,7 +17,7 @@ public Deadline(String description, String deadline) throws DukeException {
             this.date = localDate.format(DateTimeFormatter.ofPattern("MMM dd yyyy"));
         } catch (DateTimeParseException e) {
-            throw new DukeException("Deadline should be in a yyyy-mm-dd format.");
+            throw new DukeException("duke.task.Deadline should be in a yyyy-mm-dd format.");
```
You may import duke.task here.

HangZelin reviewed (commented):
Overall a pretty nice job for coding standard! And your OOP design is pretty nice too. Just to remind there can be some improvements for your import and packaging naming. 

Keep on your good job! 👍

rebeccalaujx committed:
Merge branch 'add-gradle-support' of https://github.com/rebeccalaujx/ip into branch-A-Gradle

rebeccalaujx committed:
Automate project builds using Gradle and fixed some checkstyle errors

rebeccalaujx committed:
fixed checkstyle issues

rebeccalaujx committed:
created basic GUI for Duke

rebeccalaujx committed:
used FXML for GUI

rebeccalaujx committed:
merge branch-Level-10

rebeccalaujx committed:
Ui class: cleaned up comments

rebeccalaujx committed:
Add assertions feature

The assertions feature addresses important assumptions that should
hold in the code.

A program that terminates the program early once a bug is observed
is safer to use as it prevents the error from escalating further.

Let's include the assertions feature to define assumptions about the
program state so that the runtime can verify them.

rebeccalaujx committed:
Improve code quality

Information extraction from input was largely repeated throughout
the code, creating long methods and convoluted expressions.

Extracting common behaviour into the common Input class reduces the
amount of arrowhead style code, makes the expressions easier to
understand, and shortens long methods, hence improving the code's
readability.

Let's improve the code quality so that the code can be more readable.

rebeccalaujx committed:
Merge pull request #2 from rebeccalaujx/branch-Assertions

Branch assertions

rebeccalaujx committed:
Merge pull request #1 from rebeccalaujx/branch-Assertions

rebeccalaujx committed:
Merge branch 'master' into branch-Assertions and branch-CodeQuality

rebeccalaujx committed:
Merge pull request #3 from rebeccalaujx/branch-CodeQuality

Branch code quality

rebeccalaujx committed:
Add C-MassOps

Users might want to delete many tasks or set many tasks as done
all in one go, instead of entering a new command line for each task.

MassOps allows users to delete multiple tasks or set multiple tasks
as done, all in one go.

Let's update the functions that delete tasks and set tasks as done
to allow them to perform their function on multiple tasks at once.

rebeccalaujx committed:
Add varargs

rebeccalaujx committed:
Update docs/README.md

rebeccalaujx committed:
Update docs/README.md

rebeccalaujx committed:
Update docs/README.md

rebeccalaujx committed:
Update docs/README.md

rebeccalaujx committed:
Add Ui.png

rebeccalaujx committed:
Fix Storage function

rebeccalaujx committed:
Update JavaDocs

rebeccalaujx committed:
Improve GUI

rebeccalaujx committed:
Fix Storage

rebeccalaujx committed:
change class path

rebeccalaujx committed:
fix Storage

rebeccalaujx committed:
Update Ui.png
